## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [2]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [3]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [4]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data.
var0 = df.index
print(var0)

var1 = df.columns
print(var1)

var2 = df.count()
print(var2)

var3 = df.sort_values(by='category')
print(f"this is the sroted values, {var3}")

# type = type(df)
# print(f"this the type,{type}")

RangeIndex(start=0, stop=54639, step=1)
Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')
first          54639
last           54639
job            54639
phone          54639
email          54639
client_id      54639
order_id       54639
order_date     54639
order_week     54639
order_year     54639
item_id        54639
category       54639
subcategory    54639
unit_price     54639
unit_cost      54639
unit_weight    54639
qty            54639
line_number    54639
dtype: int64
this is the sroted values,           first       last                       job                 phone  \
47373      Mark   Thompson  Journalist, broadcasting     529-839-1559x9494   
34204     Angel     Dillon         Market researcher    868.290.8408x61810   
34205   Lindsey   Robinson             Media p

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [7]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [8]:
# For the category with the most entries, which subcategory had the most entries?
df.loc[df['category']=='consumables','subcategory'].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [9]:
df.loc[df['category']=='consumables']

first       last                                      job  \
1        Tiffany      Myers                          Music therapist   
3         Nathan      Baker                    Accounting technician   
4      Christina   Schwartz                             Chiropractor   
6          Brian      Moore                                  Barista   
7          Jesus     Wright                       Theme park manager   
...          ...        ...                                      ...   
54629      Linda  Henderson  Scientist, research (physical sciences)   
54632     Austin     Fuller                       Investment analyst   
54633     Tamara      Jones                        Fish farm manager   
54635    Derrick      Moore                                 Musician   
54636     Monica  Gutierrez                         Graphic designer   

                     phone                                email  client_id  \
1         201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
3         827-788-8123x012               bakernathan@benson.com      46554   
4             265-829-3643  christinaschwartz9252@mcconnell.com      92089   
6      +1-893-400-2733x982          moore.official@martinez.com      87213   
7       605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
...                    ...                                  ...        ...   
54629  +1-664-405-4100x443         lhenderson@boyer-krueger.com      79817   
54632         379.662.7549           fulleraustin7494@nunez.com      50317   
54633     375.495.3124x931  tamara.jones.9326@george-carter.com      22109   
54635         358.661.5483         derrick.moore.2602@pope.info      41908   
54636     294.805.9100x339            gutierrezm3195@morris.org      35176   

       order_id  order_date  order_week  order_year         item_id  \
1       8069089  2023-05-19          20        2023  XDA18116-89-4A   
3       9031802  2023-04-25          17        2023  ZMM00836-65-0C   
4       1322274  2023-05-28          21        2023  BZX55559-12-3X   
6       8080280  2023-04-19          16        2023  UAC50632-25-4E   
7       8946782  2023-02-14           7        2023  UEU25316-01-1U   
...         ...         ...         ...         ...             ...   
54629   6473704  2023-03-16          11        2023  ZCC45621-74-4A   
54632   5064968  2023-05-26          21        2023  CUE68543-95-3U   
54633   8568920  2023-04-26          17        2023  CDX06274-12-1E   
54635   6290153  2023-02-08           6        2023  EBD29277-23-7U   
54636   8692622  2023-03-05           9        2023  EBD30093-55-9Z   

          category        subcategory  unit_price  unit_cost  unit_weight  \
1      consumables               pens       24.95      15.09         1.49   
3      consumables               pens       36.42      24.85         1.23   
4      consumables               misc      195.10     108.17        46.43   
6      consumables   kitchen supplies      245.56     231.95         1.73   
7      consumables               pens       20.99      16.97         5.12   
...            ...                ...         ...        ...          ...   
54629  consumables   kitchen supplies       72.60      52.42         0.33   
54632  consumables     paper products      557.31     409.72         2.21   
54633  consumables   kitchen supplies        8.16       5.10         3.84   
54635  consumables   kitchen supplies      206.59     175.46        11.70   
54636  consumables  bathroom supplies       65.66      57.31         4.16   

       qty  line_number  
1       21            0  
3       29            3  
4       20            1  
6       56            5  
7       22            4  
...    ...          ...  
54629  153            1  
54632  330            1  
54633   41            2  
54635   47            0  
54636  475            6  

[23538 rows x 18 columns]

In [10]:
# Which five clients had the most entries in the data?
df['client_id'].value_counts().head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [11]:
# Store the client ids of those top 5 clients in a list.
client_list = list(df['client_id'].value_counts().head(5).index)
client_list

[33615, 66037, 46820, 38378, 24741]

In [12]:
# How many total units (the qty column) did the client with the most entries order order?
df.loc[df['client_id']==33615,'qty'].sum()
# (df['client_id']).dtypes

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [14]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']
df[['client_id', 'qty', 'line_subtotal']]

client_id   qty  line_subtotal
0          58515   105      115164.00
1          37609    21         523.95
2          57113    39         527.28
3          46554    29        1056.18
4          92089    20        3902.00
...          ...   ...            ...
54634      29043    33        2743.29
54635      41908    47        9709.73
54636      35176   475       31188.50
54637      24485   112         165.76
54638      24007  1031        3103.31

[54639 rows x 3 columns]

In [15]:
df['total_weight'] = df['unit_weight'] * df['qty']



In [16]:
def shipping_price(total_weight):
    return 7 if total_weight>50 else 10
    
df['shipping_price'] = df['total_weight'].apply(shipping_price) * df['total_weight']
df[['unit_price','unit_weight', 'qty', 'total_weight', 'shipping_price']].head(3)

unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [17]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def calculator(total_weight):
   if total_weight > 50:
        return total_weight * 7
   else: 
        return total_weight * 10


df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df['total_weight'].apply(shipping_price) * df['total_weight']

df[['unit_price','unit_weight', 'qty', 'total_weight', 'shipping_price', 'line_subtotal']].head(3)

unit_price  unit_weight  qty  total_weight  shipping_price  line_subtotal
0     1096.80         7.50  105        787.50         5512.50      115164.00
1       24.95         1.49   21         31.29          312.90         523.95
2       13.52         1.68   39         65.52          458.64         527.28

In [18]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%

df['line_price'] = ((df['shipping_price'] + df['line_subtotal']) * 1.0925).round(2)

In [22]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
# df[['order_id'] == 2742071]

In [24]:
df[['line_subtotal', 'shipping_price', 'line_price', 'line_cost']].head(3)


line_subtotal  shipping_price  line_price  line_cost
0      115164.00         5512.50   131839.08   85597.05
1         523.95          312.90      914.26     629.79
2         527.28          458.64     1077.12     765.18

In [ ]:
# Create a column for the profit of each line using line cost and line price

df['profit'] = (df['unit_cost'] * df['qty']) + df['line_price']

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [ ]:
# Check your work using the totals above


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [ ]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.



In [ ]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 


In [ ]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 




In [ ]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
